sklearn/decomposition/_nmf.py

```diff
@@ -1097,6 +1102,15 @@ def non_negative_factorization(
     >>> W, H, n_iter = non_negative_factorization(
     ...     X, n_components=2, init='random', random_state=0)
     """
+
+    if n_components == "warn" or n_components is None:
+        warnings.warn(
+            "The default value of `n_components` will change from "
+            "`None` to `'auto'` in 1.5.",
+            FutureWarning,
+        )
+        n_components = None
```

Comment:
Done in https://github.com/scikit-learn/scikit-learn/pull/26634/commits/9f031da133b1048ca208a5191b07f65369eef6c2
I moved this to the `_check_params` function since it appears this is where the checks are done.
https://github.com/scikit-learn/scikit-learn/blob/ef6612f3b35bf2cc9c93daf8f07ebd1b3baefdc9/sklearn/decomposition/_nmf.py#L1250

Added bonus with this: the check only needs to be done in the base class `_BaseNMF` as it is called from the child classes:
- `NMF`: https://github.com/scikit-learn/scikit-learn/blob/ef6612f3b35bf2cc9c93daf8f07ebd1b3baefdc9/sklearn/decomposition/_nmf.py#L1548
- `MiniBatchNMF`: https://github.com/scikit-learn/scikit-learn/blob/ef6612f3b35bf2cc9c93daf8f07ebd1b3baefdc9/sklearn/decomposition/_nmf.py#L1992